src/java.base/share/classes/java/lang/invoke/MethodHandleProxies.java

```diff
@@ -171,64 +216,249 @@ public static <T> T asInterfaceInstance(final Class<T> intfc, final MethodHandle
-        for (int i = 0; i < methods.length; i++) {
-            Method sm = methods[i];
-            MethodType smMT = MethodType.methodType(sm.getReturnType(), sm.getParameterTypes());
-            MethodHandle checkTarget = mh.asType(smMT);  // make throw WMT
-            checkTarget = checkTarget.asType(checkTarget.type().changeReturnType(Object.class));
-            vaTargets[i] = checkTarget.asSpreader(Object[].class, smMT.parameterCount());
+
+        /*
+         * For each interface I define a new hidden class and pass the original and
+         * caller-sensitive method handles to its constructor.
+         *
+         * The hidden classes defined for I is defined in a dynamic module M
+         * which has access to the types referenced by the members of I including
+         * the parameter types, return type and exception types.
+         */
+        Lookup lookup = getProxyClassLookup(intfc);  // throws IllegalArgumentException
+        Object proxy;
+        try {
+            MethodHandle constructor = lookup.findConstructor(lookup.lookupClass(), MT_void_Lookup_MethodHandle_MethodHandle)
+                    .asType(MT_Object_Lookup_MethodHandle_MethodHandle);
+            proxy = constructor.invokeExact(lookup, target, mh);
+        } catch (Throwable ex) {
+            throw uncaughtException(ex);
         }
-        final InvocationHandler ih = new InvocationHandler() {
-                private Object getArg(String name) {
-                    if ((Object)name == "getWrapperInstanceTarget")  return target;
-                    if ((Object)name == "getWrapperInstanceType")    return intfc;
-                    throw new AssertionError();
+        assert proxy.getClass().getModule().isNamed() : proxy.getClass() + " " + proxy.getClass().getModule();
+        return intfc.cast(proxy);
+    }
+
+    private record LocalMethodInfo(MethodTypeDesc desc, List<ClassDesc> thrown, String fieldName) {}
+
+    private static final ClassFileDumper DUMPER = ClassFileDumper.getInstance(
+            "jdk.invoke.MethodHandleProxies.dumpClassFiles", "DUMP_MH_PROXY_CLASSFILES");
+
+    private static final Set<Class<?>> WRAPPER_TYPES = Collections.newSetFromMap(new WeakHashMap<>());
+    private static final ClassValue<SoftReference<Lookup>> PROXY_LOOKUPS = new ClassValue<>() {
+        @Override
+        protected SoftReference<Lookup> computeValue(Class<?> intfc) {
+            List<LocalMethodInfo> methods = new ArrayList<>();
+            Set<Class<?>> referencedTypes = new HashSet<>();
+            referencedTypes.add(intfc);
+            String uniqueName = null;
+            int count = 0;
+            for (Method m : intfc.getMethods()) {
+                if (Modifier.isStatic(m.getModifiers()))
+                    continue;
+
+                if (isObjectMethod(m))
+                    continue;
+
+                if (!Modifier.isAbstract(m.getModifiers()))
+                    continue;
+
+                String mname = m.getName();
```

Comment:
Better documentation, more comments, would be helpful.   Some suggestions below.

```suggestion
                // ensure it's SAM interface
                String mname = m.getName();
```